package robogenMessage;

message EvolvableParameter {
   required float paramValue = 1;
}

message BodyPart {
  required string id = 1;
  required string type = 2;
  required bool root = 3;
  repeated EvolvableParameter evolvableParam = 4;
  required int32 orientation = 5;
}

message BodyConnection {
  required string src = 1; 
  required string dest = 2;
  required int32 srcSlot = 3;
  required int32 destSlot = 4;
}

message Body {
  repeated BodyPart part = 1;
  repeated BodyConnection connection = 2;
}

message NeuralConnection {
  required string src = 1; 
  required string dest = 2;
  required float weight = 3;
}

message Neuron {
  required string id = 1;
  required string layer = 2;
  required string type = 3;
  optional float bias = 4;
  optional string bodyPartId = 5; 
  optional int32 ioId = 6;
  optional float tau = 7;
  optional float phaseOffset = 8;
  optional float period = 9;  
  optional float gain = 10; 
}

message Brain {
  repeated Neuron neuron = 1;
  repeated NeuralConnection connection = 2;
}

message Robot {
  required int32 id = 1;
  required Body body = 2; 
  required Brain brain = 3;
}

message Obstacle {
  required float x = 1;
  required float y = 2;
  required float z = 3;
  required float xSize = 4;
  required float ySize = 5;
  required float zSize = 6;
  required float density = 7;
  required float xRotation = 8;
  required float yRotation = 9;
  required float zRotation = 10;
  required float rotationAngle = 11;
}

message LightSource {
  required float x = 1;
  required float y = 2;
  required float z = 3;
  required float intensity = 7;
}

message StartPosition {
  required float x = 1;
  required float y = 2;
  required float azimuth = 3;
}

message SimulatorConf {
  required float terrainWidth = 1;
  required float terrainLength = 2;
  repeated Obstacle obstacles = 3;
  required string scenario = 4;
  repeated LightSource lightSources = 5;
  required float timeStep = 6;
  required int32 nTimeSteps = 7;
  repeated StartPosition startPositions = 8;
  required int32 actuationPeriod = 9;
  required float terrainFriction = 10;
  required float sensorNoiseLevel = 11;
  required float motorNoiseLevel = 12;
  required bool capAcceleration = 13;
  required float maxLinearAcceleration = 14;
  required float maxAngularAcceleration = 15;
  required int32 maxDirectionShiftsPerSecond = 16;
  required float gravityX = 17;
  required float gravityY = 18;
  required float gravityZ = 19;
  required int32 terrainType = 20;
  required float terrainHeight = 21;
  required string terrainHeightFieldFileName = 22;
  required bool disallowObstacleCollisions = 23;
  required uint32 obstacleOverlapPolicy = 24;
  
}

message EvaluationRequest {
  required Robot robot = 1;
  required SimulatorConf configuration = 2;
}

message EvaluationResult {
    required int32 id = 1; 
    required float fitness = 2;
    repeated float objectives = 3;
}

